File: AnalizadorCPP/ast.cpp
#include "ast.h"
#include <iostream>
#include "calculator.h"
#include <math.h>
// for debug information uncomment
//#define debug

AST::AST() {}

AST::~AST() {}

BinaryNode::BinaryNode(AST* left, AST* right):
   AST(),
   leftTree(left),
   rightTree(right)
{}

BinaryNode::~BinaryNode() {
#ifdef debug
   cout << "In BinaryNode destructor" << endl;
#endif

   try {
      delete leftTree;
   } catch (...) {}

   try {
      delete rightTree;
   } catch(...) {}
}

int PotenciaNode::potencia(){
   int num=getLeftSubTree()->evaluate();
   int num2=getRightSubTree()->evaluate();
   int acu=1;
   for(int i=1;i<=num2;i++){
      acu=acu*(num);
   }
   return acu;
}

AST* BinaryNode::getLeftSubTree() const {
   return leftTree;
}

AST* BinaryNode::getRightSubTree() const {
   return rightTree;
}

UnaryNode::UnaryNode(AST* sub):
   AST(),
   subTree(sub)
{}

UnaryNode::~UnaryNode() {
#ifdef debug
   cout << "In UnaryNode destructor" << endl;
#endif

   try {
      delete subTree;
   } catch (...) {}
}
   
AST* UnaryNode::getSubTree() const {
   return subTree;
}

AddNode::AddNode(AST* left, AST* right):
   BinaryNode(left,right)
{}

int AddNode::evaluate() {
   cout<<"subtree"<<endl;
   return getLeftSubTree()->evaluate() + getRightSubTree()->evaluate();
}

SubNode::SubNode(AST* left, AST* right):
   BinaryNode(left,right)
{}
TimesNode::TimesNode(AST* left, AST*right):
   BinaryNode(left,right)
{}
DivideNode::DivideNode(AST* left, AST* right):
   BinaryNode(left,right)
{}
ModuleNode::ModuleNode(AST* left, AST* right):
   BinaryNode(left,right)
{}
PotenciaNode::PotenciaNode(AST* left, AST* right):
   BinaryNode(left,right)
{}
int SubNode::evaluate() {
   cout<<"subtree"<<endl;
   return getLeftSubTree()->evaluate() - getRightSubTree()->evaluate();
}
int TimesNode::evaluate(){
   cout<<"subtree"<<endl;
   return getLeftSubTree()->evaluate()*getRightSubTree()->evaluate();
}
int DivideNode::evaluate(){
   cout<<"subtree"<<endl;
   return getLeftSubTree()->evaluate()/getRightSubTree()->evaluate();
}
int ModuleNode::evaluate(){
   cout<<"subtree"<<endl;
   return getLeftSubTree()->evaluate()%getRightSubTree()->evaluate();
}
int PotenciaNode::evaluate(){
   cout<<"subtree"<<endl;
   return pow(getLeftSubTree()->evaluate(),getRightSubTree()->evaluate());
}

NumNode::NumNode(int n) :
   AST(),
   val(n)
{}

int NumNode::evaluate() {
   return val;
}

File: AnalizadorPy1.1/Ast.py


class AST:
    def __init__(self):
        print("CONTRUCTOR AST")

    def evaluar(self) -> float:
        pass


class BinaryNode(AST):
    def __init__(self, left: AST, right: AST):
        super().__init__()
        print("CONSTRUCTOR BINARY")
        self.leftTree: AST = left
        self.rightTree: AST = right
        print("left", type(self.leftTree))
        print("right", type(self.rightTree))

    def evaluar(self) -> float:
        pass


class UnaryNode(AST):
    def __init__(self, sub: AST):
        super().__init__()
        self.subTree: AST = sub

    def evaluar(self) -> float:
        pass


class TernaryNode(AST):
    def __init__(self, left: AST, right: AST, center: AST):
        super().__init__()
        self.leftTree: AST = left
        self.rightTree: AST = right
        self.centerTree: AST = center

    def evaluar(self) -> float:
        pass


class FourNode(AST):
    def __init__(self, left: AST, right: AST, center1: AST, center2: AST):
        super().__init__()
        self.left: AST = left
        self.right: AST = right
        self.center1: AST = center1
        self.center2: AST = center2

    def evaluar(self) -> float:
        pass


class NumNode(AST):
    def __init__(self, num: float):
        super().__init__()
        self.val: float = num

    def evaluar(self) -> float:
        return self.val


class HighNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> bool:
        return self.leftTree.evaluar() > self.rightTree.evaluar()


class LowerNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> bool:
        return self.leftTree.evaluar() < self.rightTree.evaluar()


class HigherEqualNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> bool:
        return self.leftTree.evaluar() >= self.rightTree.evaluar()


class LowerEqualNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> bool:
        return self.leftTree.evaluar() <= self.rightTree.evaluar()


class EqualNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> bool:
        return self.leftTree.evaluar() == self.rightTree.evaluar()


class NotEqualNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> bool:
        return self.leftTree.evaluar() != self.rightTree.evaluar()


class AddNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        print("CONSTUCTOR ADDNODE")
        super().__init__(left, right)

    def evaluar(self) -> float:
        print("subtree")
        print("left", type(self.leftTree))
        print("right", type(self.rightTree))
        return self.leftTree.evaluar()+self.rightTree.evaluar()


class SubNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> float:
        print("subtree")
        print("left", type(self.leftTree))
        print("right", type(self.rightTree))
        return self.leftTree.evaluar()-self.rightTree.evaluar()


class TimesNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> float:
        return self.leftTree.evaluar()*self.rightTree.evaluar()


class DivideNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> float:
        return self.leftTree.evaluar()/self.rightTree.evaluar()


class ModuleNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> float:
        return self.leftTree.evaluar() % self.rightTree.evaluar()


class PowerNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> float:
        return self.leftTree.evaluar()**self.rightTree.evaluar()

#LOG REPAIR


class LogNode(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> float:
        return self.leftTree.evaluar()+self.rightTree.evaluar()


class SumCondition(BinaryNode):
    def __init__(self, left: AST, right: AST):
        super().__init__(left, right)

    def evaluar(self) -> float:
        if self.leftTree.evaluar():
            print("TRUE DE SUM CONDITION")
            return self.rightTree.evaluar()
        print("FALSE DE SUM CONDITION")
        return 1

        '''
        else:
            print("FALSE DE SUM CONDITION")
            return self.rightTree.evaluar()
        '''


class LoopOperation(FourNode):
    def __init__(self, left: AST, right: AST, center1: AST, center2: AST):
        super().__init__(left, right, center1, center2)

    #REPAIR THIS METHOD
    #def evaluate(self) -> float:
